packages/engine/lib/memory/src/shared_memory/segment.rs

```diff
@@ -1,53 +1,553 @@
-//       have batch versions, e.g. datasets.
-/// Used by datasets, agent batches, message batches, context global batch, [`PreparedBatch`].
+use std::{borrow::Borrow, env, fmt, mem, path::Path};
+
+use shared_memory::{Shmem, ShmemConf};
+use uuid::Uuid;
+
+use crate::{
+    error::{Error, Result},
+    shared_memory::{
+        padding,
+        ptr::MemoryPtr,
+        visitor::{Visit, Visitor, VisitorMut},
+        BufferChange, Metaversion,
+    },
+};
+
+/// An identifier for a shared memory section.
+#[derive(Debug, Copy, Clone, PartialEq)]
+pub struct MemoryId<'id> {
+    id: &'id Uuid,
+    suffix: u16,
+}
+
+impl<'id> MemoryId<'id> {
+    /// Creates a new identifier from the provided [`Uuid`].
+    ///
+    /// This will generate a suffix and ensures, that the shared memory segment does not already
+    /// exists at */dev/shm/*.
+    pub fn new(id: &'id Uuid) -> Self {
+        loop {
+            let memory_id = Self {
+                id,
+                suffix: rand::random::<u16>(),
+            };
+            if !Path::new(&format!("/dev/shm/{id}")).exists() {
+                return memory_id;
+            }
+        }
+    }
+
+    /// Returns the prefix used for the identifier.
+    pub fn prefix<Id: Borrow<Uuid>>(id: Id) -> String {
+        let id = id.borrow().to_simple_ref();
+        if cfg!(target_os = "macos") {
+            // MacOS shmem seems to be limited to 31 chars, probably remnants of HFS
+            // And we need to_string otherwise it's not truncated when formatting
+            format!("shm_{id:.20}")
+        } else {
+            format!("shm_{id}")
+        }
+    }
+}
+
+impl fmt::Display for MemoryId<'_> {
+    fn fmt(&self, fmt: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let prefix = Self::prefix(self);
+        if cfg!(target_os = "macos") {
+            // MacOS shmem seems to be limited to 31 chars, probably remnants of HFS
+            write!(fmt, "{}_{:.7}", prefix, self.suffix)
+        } else {
+            write!(fmt, "{}_{}", prefix, self.suffix)
+        }
+    }
+}
+
+impl Borrow<Uuid> for &MemoryId<'_> {
+    fn borrow(&self) -> &Uuid {
+        self.id
+    }
+}
+
+pub type Buffers<'a> = (&'a [u8], &'a [u8], &'a [u8], &'a [u8]);
+
+/// A memory-mapped shared memory segment.
+///
+/// Includes tools to work with internal structure.
+///
+/// # Internal Buffers
 ///
-/// [`PreparedBatch`]: crate::ffi::flush::PreparedBatch
-pub struct Segment(Memory);
+/// There are 4 main buffers contained in the shared memory which are:
+///
+///   1) Schema describing the layout of the data (could be an Arrow schema for example)
+///   2) Header data
+///   3) Meta data
+///   4) Data
+///
+/// At the beginning of the shared memory segment there is another small buffer which contains the
+/// markers to the four buffers above. This offset buffer can be read with `Memory::markers`. If one
+/// buffer is not needed, it's size can be set to `0`.
```

Comment:
I don't think this way of phrasing it is clear. I think we should make it clearer above which ones are **Optional** and then down here explain that **Optional** just means setting its size to 0.

Reply:
Technical, each one of the buffers is optional. You could even create an empty memory segment (actually, this is pointless, but you could), without violating anything.

Reply:
Are all permutations valid? Do we not have expectations that segments with a schema buffer will have a header buffer for example?

Reply:
Yes, all permutations should be valid. The calculated size also takes into account the header size (`Markers::TOTAL_MARKERS_SIZE`). As added in the TODO, I like to rework the `Buffers` representation at some point, but there should be no reason, why any of the buffers are required to have a length strictly greater than zero.

Reply:
Can we just rephrase this then to explicitly say that all buffers are optional, and can be marked as unneeded by setting their size to 0?